crates/red_knot_python_semantic/src/semantic_index/use_def/symbol_state.rs

```diff
@@ -67,227 +66,251 @@ impl ScopedDefinitionId {
         narrowing_constraints: &mut NarrowingConstraintsBuilder,
         constraint: ScopedNarrowingConstraintClause,
     ) {
-        for binding in &mut self.live_bindings {
-            binding.narrowing_constraint =
-                narrowing_constraints.add(binding.narrowing_constraint, constraint);
-        }
+        self.live_bindings =
+            symbol_states
+                .bindings
+                .map(self.live_bindings, |binding| LiveBinding {
+                    narrowing_constraint: narrowing_constraints
+                        .add(binding.narrowing_constraint, constraint),
+                    visibility_constraint: binding.visibility_constraint,
+                });
     }
 
     /// Add given visibility constraint to all live bindings.
     pub(super) fn record_visibility_constraint(
         &mut self,
+        symbol_states: &mut SymbolStatesBuilder,
         visibility_constraints: &mut VisibilityConstraintsBuilder,
         constraint: ScopedVisibilityConstraintId,
     ) {
-        for binding in &mut self.live_bindings {
-            binding.visibility_constraint = visibility_constraints
-                .add_and_constraint(binding.visibility_constraint, constraint);
-        }
+        self.live_bindings =
+            symbol_states
+                .bindings
+                .map(self.live_bindings, |binding| LiveBinding {
+                    narrowing_constraint: binding.narrowing_constraint,
+                    visibility_constraint: visibility_constraints
+                        .add_and_constraint(binding.visibility_constraint, constraint),
+                });
     }
 
     /// Iterate over currently live bindings for this symbol
-    pub(super) fn iter(&self) -> LiveBindingsIterator<'_> {
-        self.live_bindings.iter()
-    }
-
-    /// Iterate over the IDs of each currently live binding for this symbol
-    fn iter_bindings(&self) -> impl Iterator<Item = ScopedDefinitionId> + '_ {
-        self.iter().map(|lb| lb.binding)
+    pub(super) fn iter(self, symbol_states: &SymbolStatesStorage) -> LiveBindingsIterator<'_> {
+        symbol_states.bindings.iter(self.live_bindings)
     }
 
-    fn simplify_visibility_constraints(&mut self, other: SymbolBindings) {
+    fn simplify_visibility_constraints(
+        &mut self,
+        symbol_states: &mut SymbolStatesBuilder,
+        other: SymbolBindings,
+    ) {
         // If the set of live bindings hasn't changed, don't simplify.
-        if self.live_bindings.len() != other.live_bindings.len()
-            || !self.iter_bindings().eq(other.iter_bindings())
-        {
+        let self_bindings = symbol_states
+            .bindings
+            .iter(self.live_bindings)
+            .map(|(binding, _)| *binding);
+        let other_bindings = symbol_states
+            .bindings
+            .iter(other.live_bindings)
+            .map(|(binding, _)| *binding);
+        if !self_bindings.eq(other_bindings) {
```

Comment:
Is it possible to keep the `len` comparison to avoid comparing iterators of different lengths?